Generate shell completions with clap during build

davideGiovannini opened the pull request:
Auto-generate shell-completion files via clap #115

sharkdp commented on Cargo.toml:
```diff
@@ -43,3 +44,6 @@ features = []
 tempdir = "0.3"
+
+[build-dependencies]
+clap = "2.32.0"
```
I think we should use  `"2.32"` to be consistent with the dependency above.

sharkdp commented on build.rs:
```diff
@@ -0,0 +1,21 @@
+
+fn main() {
+
+    let outdir = match std::env::var_os("OUT_DIR") {
```
Can we use the same mechanism as in `fd` here:
```rust
let outdir = std::env::var_os("SHELL_COMPLETIONS_DIR").or(std::env::var_os("OUT_DIR"));
```

See https://github.com/sharkdp/fd/issues/121


It would also require us to use `fs::create_dir_all`:
https://github.com/sharkdp/fd/commit/3104729f29f4c1bb3a49b7adc2fa79f355115fd8

sharkdp commented on ci/before_deploy.bash:
```diff
@@ -38,6 +38,11 @@ pack() {
     cp LICENSE-APACHE "$tempdir/$package_name"
 
+    # various autocomplete
+    cp target/"$TARGET"/release/build/"$PROJECT_NAME"-*/out/"$PROJECT_NAME".bash "$tempdir/$package_name/autocomplete/${PROJECT_NAME}.bash-completion"
```
We need to make sure that the target directory exists.

sharkdp replied:
Also, it would be great to also do this for Windows deployments (in the `before_deploy.ps1` script). I think we don't need a subfolder in case of the Windows archive.

davideGiovannini replied:
I'm not familiar with how it should work on Windows. 

Should it be something like this ?
```ps1
Copy-Item "$SRC_DIR\target\release\build\bat-*/out/_bat.ps1" '.\autocomplete'
```

davideGiovannini commented:
>You can cargo +nightly install rustfmt-nightly and then run cargo fmt from the repository folder. 

is this the same as running `cargo fmt` installed with `rustup component add rustfmt-preview` ?
It's changing the formatting of files that I did not change 
```
        modified:   src/assets.rs
        modified:   src/diff.rs
        modified:   src/printer.rs
```
> Do wildcards work inside quotes?

Uhm probably not.
I'm removing the quotes

sharkdp commented:
Hm, I'm not sure, but there might very well be a change in the format between this and rustfmt-nightly (which we currently use to enforce the style guide). This is unfortunate, but apparently there will be a 1.0 release soon which should stabilize the format.

davideGiovannini commented:
>You can cargo +nightly install rustfmt-nightly and then run cargo fmt from the repository folder. 

is this the same as running `cargo fmt` installed with `rustup component add rustfmt-preview` ?
It's changing the formatting of files that I did not change 
```
        modified:   src/assets.rs
        modified:   src/diff.rs
        modified:   src/printer.rs
```
> Do wildcards work inside quotes?

Uhm probably not.
I'm removing the quotes

sharkdp commented:
Thank you very much!!
